google/cloud/security/scanner/audit/bigquery_rules_engine.py

```diff
@@ -0,0 +1,257 @@
+                        rule_index=rule_index,
+                        rules=rule_def_resource)
+
+            if not self.resource_rules_map.get(rule_index):
+                self.resource_rules_map[rule_index] = rule
+
+    def get_resource_rules(self):
+        """Get all the resource rules for (resource, RuleAppliesTo.*).
+
+        Returns:
+            list: A list of ResourceRules.
+        """
+        resource_rules = []
+
+        for resource_rule in self.resource_rules_map:
+            resource_rules.append(self.resource_rules_map[resource_rule])
+
+        return resource_rules
+
+
+class Rule(object):
+    """Rule properties from the rule definition file.
+       Also finds violations.
+    """
+
+    rule_violation_attributes = ['resource_type', 'resource_id', 'rule_name',
+                                 'rule_index', 'violation_type', 'dataset_id',
+                                 'role', 'special_group', 'user_email',
+                                 'domain', 'group_email']
+    frozen_rule_attributes = frozenset(rule_violation_attributes)
+    RuleViolation = namedtuple(
+        'RuleViolation',
+        frozen_rule_attributes)
+
+    def __init__(self, rule_name, rule_index, rules):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule.
+            rule_index (int): The index of the rule from the rule definitions.
+            rules (dict): The rules from the file.
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rules = rules
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_policy_violations(self, bigquery_acl):
+        """Find BigQuery acl violations in the rule book.
+
+        Args:
+            bigquery_acl (BigqueryAccessControls): BigQuery ACL resource.
+
+        Yields:
+            namedtuple: Returns RuleViolation named tuple.
+        """
+        is_dataset_id_violated = True
+        is_special_group_violated = True
+        is_user_email_bool_violated = True
+        is_domain_violated = True
+        is_group_email_violated = True
+        is_role_violated = True
+
+        is_dataset_id_violated = re.match(self.rules.dataset_id,
+                                          bigquery_acl.dataset_id)
+
+        is_special_group_violated = re.match(self.rules.special_group,
+                                             bigquery_acl.special_group)
+
+        is_user_email_bool_violated = re.match(self.rules.user_email,
+                                               bigquery_acl.user_email)
+
+        is_domain_violated = re.match(self.rules.domain, bigquery_acl.domain)
+
+        is_group_email_violated = re.match(self.rules.group_email,
+                                           bigquery_acl.group_email)
+
+        is_role_violated = re.match(self.rules.role, bigquery_acl.role)
+
+        should_raise_violation = all([
```

Comment:
By using all() here, the bigquery rules are meant to be a blacklist?  If so, can we document this?  Perhaps in the yaml file?  This would help users a lot.  :)

Maybe something like this:
```yaml
# The bigquery scanner rules are blacklists.
```